packages/api/cms-api/src/user-permissions/auth/user-permissions.guard.spec.ts

```diff
@@ -0,0 +1,370 @@
+
+@Entity()
+class TestEntity extends BaseEntity<TestEntity, "id"> {
+    @PrimaryKey()
+    id: number;
+}
+
+class AccessControlService extends AbstractAccessControlService {}
+
+describe("UserPermissionsGuard", () => {
+    let guard: UserPermissionsGuard;
+    let reflector: Reflector;
+    let orm: MikroORM;
+    let contentScopeService: ContentScopeService;
+    let accessControlService: AccessControlService;
+
+    const mockAnnotations = (annotations: {
+        requiredPermission?: RequiredPermissionMetadata;
+        affectedEntities?: AffectedEntityMeta[];
+        scopedEntity?: ScopedEntityMeta["fn"];
+    }) => {
+        reflector.getAllAndOverride = jest.fn().mockImplementation((decorator: string) => {
+            if (decorator === "requiredPermission") return annotations.requiredPermission;
+            if (decorator === "affectedEntities") return annotations.affectedEntities;
+            if (decorator === "scopedEntity") return { fn: annotations.scopedEntity };
+            return false;
+        });
+    };
+    const mockContext = (context: { userPermissions: CurrentUser["permissions"]; args?: unknown }) => {
+        return createMock<ExecutionContext>({
+            switchToHttp: () => ({
+                getRequest: () => ({
+                    user: {
+                        id: "1",
+                        name: "Admin",
+                        email: "demo@comet-dxp.com",
+                        language: "en",
+                        permissions: context.userPermissions,
+                    } satisfies CurrentUser,
+                    params: context.args,
+                }),
+            }),
+        });
+    };
+    const mockAffectedEntityValues = (values: { id: number; [key: string]: unknown }[]) => {
+        orm.em.getRepository = jest
+            .fn()
+            .mockReturnValue({ findOneOrFail: jest.fn().mockImplementation((id: number) => values.find((v) => v.id === id)) });
+    };
+
+    beforeEach(async () => {
+        reflector = new Reflector();
+        orm = await MikroORM.init({
+            type: "postgresql",
+            dbName: "test-db",
+            entities: [TestEntity],
+            connect: false,
+            allowGlobalContext: true,
+        });
+        contentScopeService = new ContentScopeService(reflector, orm);
+        accessControlService = new AccessControlService();
+        guard = new UserPermissionsGuard(reflector, contentScopeService, accessControlService);
+    });
+
+    test("basic permission check", async () => {
+        mockAnnotations({
+            requiredPermission: {
+                requiredPermission: ["p1"],
+                options: { skipScopeCheck: true },
+            },
+        });
+
+        expect(
+            await guard.canActivate(
+                mockContext({
+                    userPermissions: [{ permission: "p1", contentScopes: [] }],
+                }),
+            ),
+        ).toBe(true);
```

Comment:
What do you think about moving each expect into a separate test? The test description could then provide some context which branch you're testing.

Reply:
It's a lot of mocking, but I guess that's inevitable, so the approach looks good to me.